gdnative-core/src/core_types/variant.rs

```diff
@@ -1881,6 +1883,294 @@ macro_rules! impl_variant_for_tuples {
+
+    impl Serialize for Variant {
+        #[inline]
+        fn serialize<S>(
+            &self,
+            serializer: S,
+        ) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+        where
+            S: Serializer,
+        {
+            serializer.serialize_newtype_struct("Variant", &self.dispatch())
+        }
+    }
+
+    struct VariantDispatchVisitor;
+
+    impl<'de> Visitor<'de> for VariantDispatchVisitor {
+        type Value = VariantDispatch;
+
+        fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
+            formatter.write_str("a Variant")
+        }
+
+        fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            Ok(VariantDispatch::Bool(v))
+        }
+
+        fn visit_i8<E>(self, v: i8) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_i16<E>(self, v: i16) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_i32<E>(self, v: i32) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            Ok(VariantDispatch::I64(v))
+        }
+
+        fn visit_i128<E>(self, v: i128) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_u8<E>(self, v: u8) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_u16<E>(self, v: u16) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_u32<E>(self, v: u32) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_u128<E>(self, v: u128) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_i64(v as i64)
+        }
+
+        fn visit_f32<E>(self, v: f32) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            self.visit_f64(v as f64)
+        }
+
+        fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            Ok(VariantDispatch::F64(v))
+        }
+
+        fn visit_char<E>(self, _v: char) -> Result<Self::Value, E>
+        where
+            E: Error,
+        {
+            Err(E::custom("TODO: decide how to deserialize char to Variant"))
```

Comment:
String?

Reply:
I see 3 options:

- As you said, just a string of length 1
    - Then would all String variants that happen to be length 1 be serialized as a `char`?
- A `u32` or `[u8; 4]` [(`char` is always 4 bytes in size)](https://doc.rust-lang.org/std/primitive.char.html#representation)
    - Similar issue, how do we know when to treat a variant as a `char` vs a number or `ByteArray`?
- Tagged somehow, as a struct or map, e.g. `Char { value: u32 }`
    - Seems quite overkill and probably surprising for it to take up so much space

We should probably also implement `ToVariant/FromVariant` for `char` in the same way, however we do it.

Reply:
We should just never serialize things into `char`. It isn't present in many of the popular serialization formats, and it also isn't in Godot's data model. FYI, `visit_char` forwards to `visit_str` by default: https://docs.serde.rs/serde/de/trait.Visitor.html#method.visit_char